### PR TITLE
SRS-372: Suspect Land Uses Tab

### PR DESCRIPTION
**Pull request by anton-bcgov**
[SRS-372: Suspect Land Uses Tab](https://epddigitalservices.atlassian.net/browse/SRS-372)

Screenshot 2024-07-25 at 9 07 27 AM

### CODE REVIEW

File: backend/sites/src/app/services/landHistory/landHistory.service.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+export class LandHistoryService {
+    constructor(
+        @InjectRepository(LandHistories)
+        private landHistoryRepository: Repository<LandHistories>,
+    ) { }
+
+    async getLandHistoriesForSite(siteId: string): Promise<LandHistories[]> {
+        try {
+            return this.landHistoryRepository.find({ where: { siteId } });
```
Comment (nikhila-aot) [inline backend/sites/src/app/services/landHistory/landHistory.service.ts]: assign this to a variable and return. Wil need to be used for logging in future.

File: backend/sites/src/app/resolvers/landHistory/landHistory.resolver.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        private readonly genericResponseProvider: GenericResponseProvider<LandHistories[]>,
+    ) { }
+
+    @Roles({ roles: ['site-admin'], mode: RoleMatchingMode.ANY })
+    @Query(() => LandHistoryResponse, { name: 'getLandHistoriesForSite' })
+    @UsePipes(new GenericValidationPipe())
+    async getLandHistoriesForSite(
+        @Args('siteId', { type: () => String }) siteId: string,
+    ) {
+        const result = await this.landHistoryService.getLandHistoriesForSite(siteId);
```
Comment (nikhila-aot) [inline backend/sites/src/app/resolvers/landHistory/landHistory.resolver.ts]: handle the case when result does not contain data

File: backend/sites/src/app/entities/landUseCd.entity.ts [hunk 3]
```diff
@@ -16,7 +16,7 @@ export class LandUseCd {
   @Column("character varying", { name: "description", length: 60 })
   description: string;
 
-  @OneToMany(() => LandHistories, (landHistories) => landHistories.lutCode2)
+  @OneToMany(() => LandHistories, (landHistories) => landHistories.landUse)
```
Comment (nikhila-aot) [inline backend/sites/src/app/entities/landUseCd.entity.ts:18]: isn't this lutCode2?

File: backend/sites/src/app/entities/landHistories.entity.ts [hunk 4]
```diff
@@ -69,6 +69,11 @@ export class LandHistories {
   @JoinColumn([{ name: "lut_code", referencedColumnName: "code" }])
   lutCode2: LandUseCd;
 
+  @Field(() => LandUseCd)
```
Comment (nikhila-aot) [inline backend/sites/src/app/entities/landHistories.entity.ts:68]: there is already a relationship existing in line 68-70. Can we use the same?

File: backend/sites/src/app/services/landHistory/landHistory.service.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+        } catch (error) {
+            throw error;
+        }
+    }
+}
```
Comment (nikhila-aot) [inline backend/sites/src/app/services/landHistory/landHistory.service.ts:48]: pls include unit tests for resolver and services

**Review by nikhila-aot — CHANGES_REQUESTED** [review]

File: backend/sites/src/app/resolvers/landHistory/landHistory.resolver.ts (hunk 2 above)
Comment (anton-bcgov) [reply backend/sites/src/app/resolvers/landHistory/landHistory.resolver.ts]: Could you elaborate? We return an empty array when there's no data found.

File: backend/sites/src/app/entities/landHistories.entity.ts (hunk 4 above)
Comment (anton-bcgov) [reply backend/sites/src/app/entities/landHistories.entity.ts:68]: yeah, I was going to ask you about this. I think `lutCode2` is not a very descriptive name, but I wasn't 100% confident in removing that relationship altogether, so I created a new one. What do you think? Should we leave `lutCode2` as is, remove it, or straight up use it?

File: backend/sites/src/app/entities/landUseCd.entity.ts (hunk 3 above)
Comment (anton-bcgov) [reply backend/sites/src/app/entities/landUseCd.entity.ts:18]: discussing this in another thread

File: backend/sites/src/app/services/landHistory/landHistory.service.ts (hunk 5 above)
Comment (anton-bcgov) [reply backend/sites/src/app/services/landHistory/landHistory.service.ts:48]: That's the plan! I didn't expect you to review a Draft PR 😄

**Review by midhun-aot — COMMENTED** [review]
Frontend changes looks good.

File: backend/sites/src/app/entities/landHistories.entity.ts (hunk 4 above)
Comment (nikhila-aot) [reply backend/sites/src/app/entities/landHistories.entity.ts:68]: we have to leave it intact, we cannot change any existing column names as they are replicated from the oracle db.

File: backend/sites/src/app/services/landHistory/landHistory.service.ts (hunk 5 above)
Comment (nikhila-aot) [reply backend/sites/src/app/services/landHistory/landHistory.service.ts:48]: the team raises the PR when the task is done, not before that 😄

File: backend/sites/src/app/resolvers/landHistory/landHistory.resolver.ts (hunk 2 above)
Comment (nikhila-aot) [reply backend/sites/src/app/resolvers/landHistory/landHistory.resolver.ts]: example:

if (result.length > 0) {
          return this.genericResponseProvider.createResponse('Participants fetched successfully', 200, true, result);
        }
        else
        {
          return this.genericResponseProvider.createResponse(`Participants data not found for site id: ${siteId}`, 404, false);
        }

File: backend/sites/src/app/entities/landHistories.entity.ts (hunk 4 above)
Comment (anton-bcgov) [reply backend/sites/src/app/entities/landHistories.entity.ts:68]: It's not changing the database columns though, right? This only defines a relationship in the application code, the underlying data still comes from the `lut_code` and `code` DB columns,

Comment (nikhila-aot) [reply backend/sites/src/app/entities/landHistories.entity.ts:68]: its redundant code, for the same purpose. Just to have a better column name

Comment (anton-bcgov) [reply backend/sites/src/app/entities/landHistories.entity.ts:68]: just so we're on the same page, do you agree that we can remove `lutCode2` relationship if it's safe to do so and prefer a clearer variable name?

**Comment by anton-bcgov** [comment]
Sorry for the unrelated formatting changes that are included in this PR. For some reason, formatter didn't run on #23, and ran instead when I did a merge commit from `dev` branch :shrug:

File: frontend/site-registry/src/app/components/table/body/TableBody.tsx [hunk 6]
```diff
@@ -118,184 +119,191 @@ const TableBody: FC<TableBodyProps> = ({
...
+          tableMode={field.tableMode ?? false}
+          stickyCol={field.stickyCol}
+        />
+      );
+    }
   };
 
   const getValue = (rowIndex: number, propertyName: string) => {
-    return data[rowIndex][propertyName];
+    return get(data[rowIndex], propertyName);
```
Comment (midhun-aot) [inline frontend/site-registry/src/app/components/table/body/TableBody.tsx:392]: Please remove the import from loadash as we will introducing an external dependency. May be to check for undefined or null we can use (data[rowIndex][propertyName]) ?? ''

File: backend/sites/src/app/resolvers/landHistory/landHistory.resolver.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  constructor(
+    private readonly landHistoryService: LandHistoryService,
+    private readonly genericResponseProvider: GenericResponseProvider<
+      LandHistories[]
+    >,
+  ) {}
+
+  @Roles({ roles: ['site-admin'], mode: RoleMatchingMode.ANY })
+  @Query(() => LandHistoryResponse, { name: 'getLandHistoriesForSite' })
+  // TODO: figure out how this works and if we need GenericValidationPipe here
```
Comment (nikhila-aot) [inline backend/sites/src/app/resolvers/landHistory/landHistory.resolver.ts]: is there more logic to be added here ?

File: backend/sites/src/app/resolvers/landHistory/landHistory.resolver.ts [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+    @Args('sortDirection', { nullable: true })
+    sortDirection: SortDirection,
+  ) {
+    const result = await this.landHistoryService.getLandHistoriesForSite(
+      siteId,
+      searchTerm,
+      sortDirection,
+    );
+    if (result.length > 0) {
```
Comment (nikhila-aot) [inline backend/sites/src/app/resolvers/landHistory/landHistory.resolver.ts:36]: should you check if result is empty before performing result.length ?

File: backend/sites/src/app/entities/landHistories.entity.ts (hunk 4 above)
Comment (nikhila-aot) [reply backend/sites/src/app/entities/landHistories.entity.ts:68]: yes, keep the one with the new name and remove the old one, we do not have to retain redundant relationships.

File: backend/sites/src/app/resolvers/landHistory/landHistory.resolver.ts (hunk 8 above)
Comment (anton-bcgov) [reply backend/sites/src/app/resolvers/landHistory/landHistory.resolver.ts:36]: `getLandHistoriesForSite` calls typeorm's `getMany`, which should always return an array, so I think we're safe here.

File: backend/sites/src/app/resolvers/landHistory/landHistory.resolver.ts (hunk 7 above)
Comment (anton-bcgov) [reply backend/sites/src/app/resolvers/landHistory/landHistory.resolver.ts]: nope, it's an old comment, I deleted it

File: backend/sites/src/app/entities/landHistories.entity.ts (hunk 4 above)
Comment (anton-bcgov) [reply backend/sites/src/app/entities/landHistories.entity.ts:68]: makes sense. done!

**Review by midhun-aot — APPROVED** [review]

**Review by midhun-aot — APPROVED** [review]

**Review by nikhila-aot — APPROVED** [review]

**Pull request merged**